src/main/java/io/confluent/copycat/hdfs/WAL.java

```diff
@@ -0,0 +1,161 @@
+  private WALFile.Reader reader = null;
+  private FileSystem fs = null;
+  private String logFile = null;
+  private Configuration conf = null;
+  private long sleepIntervalMillis = 1000L;
+  private int num_retry = 0;
+
+  public WAL(Configuration conf, String url, String topicsDir, TopicPartition topicPart)
+      throws CopycatException {
+    try {
+      this.conf = conf;
+      fs = FileSystem.newInstance(URI.create(url), conf);
+      logFile = FileUtils.logFileName(url, topicsDir, topicPart);
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  public void append(String tempFile, String committedFile) throws CopycatException {
+    try {
+      acquireLease();
+      WALEntry key = new WALEntry(tempFile);
+      WALEntry value = new WALEntry(committedFile);
+      writer.append(key, value);
+      writer.hsync();
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  private void acquireLease() throws CopycatException {
+    long initialSleepIntervalMillis = sleepIntervalMillis;
+    while (sleepIntervalMillis < MAX_SLEEP_INERVAL_MILLIS) {
+      try {
+        if (writer == null) {
+          writer = WALFile.createWriter(conf,
+                                        Writer.file(new Path(logFile)),
+                                        Writer.appendIfExists(true));
+        }
+        log.info("Successfully acquired lease for {}", logFile);
+        break;
+      } catch (RemoteException e) {
+        if (e.getClassName().equals(leaseException)) {
+          sleepIntervalMillis = sleepIntervalMillis * 2;
+          log.info("Cannot acquire lease on WAL {}", logFile);
+          try {
+            Thread.sleep(sleepIntervalMillis);
+          } catch (InterruptedException ie) {
+            throw new CopycatException(ie);
+          }
+        } else {
+          throw new CopycatException(e);
+        }
+      } catch (IOException e) {
+        throw new CopycatException("Error creating writer for log file " + logFile, e);
+      }
+    }
+    if (sleepIntervalMillis >= MAX_SLEEP_INERVAL_MILLIS) {
+      if (num_retry <= MAX_NUM_RETRY) {
+        num_retry++;
+        throw new TransientException("Cannot acquire lease after timeout, will retry.");
+      } else {
+        throw new CopycatException("Cannot acquire lease after retry.");
+      }
+    }
+    sleepIntervalMillis = initialSleepIntervalMillis;
+  }
+
+  public void apply() throws CopycatException {
+    try {
+      if (!fs.exists(new Path(logFile))) {
+        return;
+      }
+      acquireLease();
+      if (reader == null) {
+        reader = new WALFile.Reader(conf, Reader.file(new Path(logFile)));
+      }
+      WALEntry key = new WALEntry();
+      WALEntry value = new WALEntry();
+      while (reader.next(key, value)) {
+        String tempFile = key.getFilename();
+        String committedFile = value.getFilename();
+        Path tempPath = new Path(tempFile);
+        if (fs.exists(tempPath)) {
```

Comment:
I think there's a race here between the `exists` and `renameFile` calls. Under normal circumstances, only the WAL owner should be making modifications here. But if a previous WAL owner is making modifications just after its lease expired (limiting actions outside the WAL file based on the WAL lease is purely advisory), then we may see the file existing but then it could have been moved. 

Does this code handle that case properly? Do we even need to check the existence of `tempPath`? Could we instead focus on the existence of `committedFile`? I think we could loop while the `committedFile` is missing, and in that loop try to move the `tempFile` to the `committedFile`. We might want to check if `tempFile` *doesn't* exist since then the `renameFile` operation would obviously fail, but other than that I think we want to check the `committedFile` to determine whether this operation is complete.